pkg/mirror/mirror.go

```diff
@@ -0,0 +1,116 @@
+package mirror
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	"k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/operator-framework/operator-registry/pkg/sqlite"
+)
+
+type Mirrorer interface {
+	Mirror() (map[string]string, error)
+}
+
+// DatabaseExtractor knows how to pull an index image and extract its database
+type DatabaseExtractor interface {
+	Extract(from string) (string, error)
+}
+
+type DatabaseExtractorFunc func(from string) (string, error)
+
+func (f DatabaseExtractorFunc) Extract(from string) (string, error) {
+	return f(from)
+}
+
+// ImageMirrorer knows how to mirror an image from one registry to another
+type ImageMirrorer interface {
+	Mirror(mapping map[string]string) error
+}
+
+type ImageMirrorerFunc func(mapping map[string]string) error
+
+func (f ImageMirrorerFunc) Mirror(mapping map[string]string) error {
+	return f(mapping)
+}
+
+type IndexImageMirrorer struct {
+	ImageMirrorer       ImageMirrorer
+	DatabaseExtractor   DatabaseExtractor
+
+	// options
+	Source, Dest        string
+}
+
+var _ Mirrorer = &IndexImageMirrorer{}
+
+func NewIndexImageMirror(config *IndexImageMirrorerOptions, options ...ImageIndexMirrorOption) (*IndexImageMirrorer, error) {
```

Comment:
see [comment](https://github.com/operator-framework/operator-registry/pull/https://github.com/operator-framework/operator-registry/blob/cfc214de5a0ee31a76f32060a97d3175d606fdfe/pkg/appregistry/builder.go#r339886519) for this